src/hotspot/share/opto/loopPredicate.cpp

```diff
@@ -209,9 +238,72 @@ ProjNode* PhaseIdealLoop::create_new_if_for_predicate(ProjNode* cont_proj, Node*
   return if_cont->as_Proj();
 }
 
+// Clone data nodes for the fast loop while creating a new If with create_new_if_for_predicate. Returns the node which is
+// used for the uncommon trap phi input.
+Node* PhaseIdealLoop::clone_data_nodes_for_fast_loop(Node* phi_input, ProjNode* uncommon_proj, Node* if_uct, Node_List* old_new) {
+  // Step 1: Clone all nodes on the data chain but do not rewire anything, yet. Keep track of the cloned nodes
+  // by using the old_new mapping. This mapping is then used in step 2 to rewire the cloned nodes accordingly.
+  uint last_idx = C->unique();
+  Unique_Node_List list;
+  list.push(phi_input);
+  for (uint j = 0; j < list.size(); j++) {
+    Node* next = list.at(j);
+    Node* clone = next->clone();
+    _igvn.register_new_node_with_optimizer(clone);
+    old_new->map(next->_idx, clone);
+    if (next->in(0) == uncommon_proj) {
+      // Found the last data node in the chain. Do not break, because we might have some unprocessed nodes
+      // on some path.
+      continue;
+    }
+    for (uint k = 1; k < next->req(); k++) {
+      Node* in = next->in(k);
+      if (!in->is_Phi() && get_ctrl(in) == uncommon_proj) {
+        list.push(in);
+      }
+    }
+  }
+
+  // Step 2: All nodes are cloned. Rewire them by using the old_new mapping.
+  for (uint j = 0; j < list.size(); j++) {
+    Node* next = list.at(j);
+    Node* clone = old_new->at(next->_idx);
+    assert(clone != NULL && clone->_idx >= last_idx, "must exist and be a proper clone");
+    if (next->in(0) == uncommon_proj) {
+      // For the last data node in the chain, we only need to update the control input to the new uncommon trap projection.
+      _igvn.replace_input_of(clone, 0, if_uct);
+      set_ctrl(clone, if_uct);
+    } else {
+      // All other nodes require a rewiring to the cloned input nodes.
+      for (uint k = 1; k < next->req(); k++) {
+        Node* in = next->in(k);
+        if (!in->is_Phi()) {
+          assert(!in->is_CFG(), "must be data node");
+          Node* in_clone = old_new->at(in->_idx);
+          if (in_clone != NULL && in_clone->_idx >= last_idx) {
```

Comment:
Doesn't in_clone != NULL implies in_clone->_idx >= last_idx? Should it be an assert?

Reply:
I'm not sure because `old_new` is shared in loopopts. I was concerned that I find old mappings here which I tried to exclude.

Reply:
Shouldn't it be reset between use i.e.:
memset( worklist.adr(), 0, worklist.Size()*sizeof(Node*) );
?

Reply:
Yes, I was mainly worried about visiting nodes from the current unswitching phase. But I think that cannot happen. I'll change it into an assert.